ocw/lib/openstack.py

```diff
@@ -0,0 +1,108 @@
+from datetime import datetime, timezone
+from typing import Dict
+from dateutil.parser import parse
+import openstack
+from openstack.exceptions import OpenStackCloudException
+from webui.PCWConfig import PCWConfig
+from .provider import Provider
+
+
+class Openstack(Provider):
+    __instances: Dict[str, "Openstack"] = {}
+
+    def __init__(self, namespace: str):
+        super().__init__(namespace)
+        self.client()
+
+    def __new__(cls, vault_namespace: str):
+        if vault_namespace not in Openstack.__instances:
+            Openstack.__instances[vault_namespace] = self = object.__new__(cls)
+            self.__client = None
+        return Openstack.__instances[vault_namespace]
+
+    def client(self) -> None:
+        if self.__client is None:
+            self.__client = openstack.connect(
+                insecure=True,  # Trust the certificate
+                auth_url=self.get_data('auth_url'),
+                project_name=self.get_data('project_name'),
+                username=self.get_data('username'),
+                password=self.get_data('password'),
+                region_name=self.get_data('region_name'),
+                user_domain_name=self.get_data('user_domain_name'),
+                project_domain_name=self.get_data('project_domain_id'),
+            )
+        return self.__client
+
+    def cleanup_all(self) -> None:
+        self._cleanup_instances()
+        self._cleanup_images()
+        self._cleanup_keypairs()
+
+    def _cleanup_instances(self) -> None:
+        # Delete VM's & associated floating IP address(es)
+        try:
+            servers = [vm for vm in self.client().compute.servers() if vm.name.startswith("openqa-vm-")]
```

Comment:
do we really need this kind of limitation ? why we want to limit our self's here ?

Reply:
why deletion of VM's is not obeying `dry_run` option ?

Reply:
we not checking age of VM is it intentional ?

Reply:
sorry I still not fully functioning >_< ofc. you just collecting servers and not actually deleting them so please ignore second and 3d questions

Reply:
Our VM's start with "openqa-vm-".  Do you want to make it configurable?  We _have_ to limit ourselves here because we only want to destroy VM's created by openQA.

Reply:
why ? isn't this cluster used ONLY for openQA testing ? who else using it and for what ?  From experience which I have for cleanup VMs in Azure/AWS/GCE . At the beginning we also used this super safe approach where we delete only VMs with certain pattern in the name but than it appears that we have sooo many cases where because of this I need to manually cleanup VMs which name is not following this pattern because of the "reasons".  So after all I decided to drop this extra filter because anyway 99% VMs which are created in our accounts suppose to be cleaned up and for 1% we have "pcw_ignore"  tag which can protect them .

Reply:
These non-openQA VM's are running right now:

`['fgerling-ankh', 'fgerling-retrospected', 'SLE 15-SP3 GMC', 'yifan0']`

I don't think it's wise to push the `pcw_ignore` tag for Openstack users.  We only pay for the electricity here, so this safe approach makes sense, IMO.